pkg/cmd/label/clone.go

```diff
@@ -0,0 +1,141 @@
+	"github.com/cli/cli/v2/utils"
+	"github.com/spf13/cobra"
+)
+
+type cloneOptions struct {
+	BaseRepo   func() (ghrepo.Interface, error)
+	HttpClient func() (*http.Client, error)
+	IO         *iostreams.IOStreams
+
+	SourceRepo ghrepo.Interface
+	Force      bool
+}
+
+func newCmdClone(f *cmdutil.Factory, runF func(*cloneOptions) error) *cobra.Command {
+	opts := cloneOptions{
+		HttpClient: f.HttpClient,
+		IO:         f.IOStreams,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "clone <source [HOST/]OWNER/REPO>",
+		Short: "Clones labels from another repo",
+		Long: heredoc.Doc(`
+			Clones labels from another repo into the current repo on GitHub.
+
+			All labels from the source repo will be copied to the current repo.
+			Any labels in the current repo that are not in the source repo will
+			not be deleted or modified.
+
+			If any labels from the source repo already exist in the current repo
+			the command will fail. You can overwrite existing labels in the
+			current repo using the --force flag.
+
+		`),
+		Example: heredoc.Doc(`
+			# clone and overwrite labels from the cli/cli repo into the current repo
+			$ gh label clone cli/cli --force
+		`),
+		Args: cmdutil.ExactArgs(1, "cannot clone labels: source repo argument required"),
+		RunE: func(c *cobra.Command, args []string) error {
+			var err error
+			opts.BaseRepo = f.BaseRepo
+			opts.SourceRepo, err = ghrepo.FromFullName(args[0])
+			if err != nil {
+				return err
+			}
+			if runF != nil {
+				return runF(&opts)
+			}
+			return cloneRun(&opts)
+		},
+	}
+
+	cmd.Flags().BoolVarP(&opts.Force, "force", "f", false, "Overwrite labels in the current repo from the source repo.")
+
+	return cmd
+}
+
+func cloneRun(opts *cloneOptions) error {
+	baseRepo, err := opts.BaseRepo()
+	if err != nil {
+		return err
+	}
+
+	successCount, totalCount, err := cloneLabels(opts)
+	if err != nil {
+		return err
+	}
+
+	if opts.IO.IsStdoutTTY() {
+		cs := opts.IO.ColorScheme()
+		pluralize := func(num int) string {
+			return utils.Pluralize(num, "label")
+		}
+
+		switch {
+		case successCount == totalCount:
+			fmt.Fprintf(opts.IO.Out, "%s Cloned %s from %s to %s\n", cs.SuccessIcon(), pluralize(successCount), opts.SourceRepo, baseRepo)
+		case successCount != 0:
+			fmt.Fprintf(opts.IO.Out, "%s Cloned %s of %d from %s to %s\n", cs.WarningIcon(), pluralize(successCount), totalCount, opts.SourceRepo, baseRepo)
+		default:
+			fmt.Fprintf(opts.IO.Out, "%s Failed to clone %s from %s to %s\n", cs.FailureIcon(), pluralize(totalCount), opts.SourceRepo, baseRepo)
+		}
+	}
+
+	return nil
+}
+
+func cloneLabels(opts *cloneOptions) (successCount, totalCount int, err error) {
+	httpClient, err := opts.HttpClient()
+	if err != nil {
+		return
+	}
+
+	baseRepo, err := opts.BaseRepo()
+	if err != nil {
+		return
+	}
+
+	opts.IO.StartProgressIndicator()
+	defer opts.IO.StopProgressIndicator()
+
+	successCount = 0
+	labels, totalCount, err := listLabels(httpClient, opts.SourceRepo, listLimit)
```

Comment:
What happens if there are more than the `listLimit` number of labels? Seems like we should check first how many labels there are and then use that here rather than hardcode the `listLimit`.

Reply:
I admit I didn't scan through enough of `listLabels`. Initially it seemed like the limit was just a page size. I see now it's actually a hard limit. IMO, this shouldn't be in `label list`. Why not just list all labels using pageable results like I do in heaths/gh-label? I don't see a customer value here, especially labels are not returned in any documented deterministic order.

IMO, we should actually eliminate the default value of 30 and not use a limit if not specified explicitly. I.e. page from `listLabels`. The limit could be made optional - if not in the GraphQL query (if supported), we can use separate GraphQL queries for limited vs. paged.

Reply:
To clarify, it is used as the page size and does support paging, but lines 67-69 of this block of code should be removed:

https://github.com/cli/cli/blob/c99c47f34db5d1ff742ae2bb89a64d7b6fb2aa06/pkg/cmd/label/list/http.go#L65-L70

Happy to do that in this PR or another, but I don't see any customer value in this. For issues, a limit makes sense because there are additional query parameters you can specify. Nothing like that exists for labels, but maybe it should. My [extension](https://github.com/heaths/gh-label#list) does for use with a repo that has a lot of labels e.g. [Azure/azure-sdk-for-net](https://github.com/Azure/azure-sdk-for-net). Even then, seems a limit should not be used by default, especially given the non-deterministically unordered results.

Reply:
Opened #5489 to track.

Reply:
IMO the customer value provided by having the limit is two fold. 

First the UX is better with shorter lists on the terminal, 30 items generally fit on a single terminal screen and doesn't take up too much terminal height. It is also consistent with our other list commands which I believe use 30 as a default. The UX argument is subjective, and can be combated by starting up the pager process when outputting the label list. 

The second, is that it keeps the user from potentially burning through their API rate limit inadvertently. Likely, the couple extra requests needed in most cased to retrieve all the repo's labels won't greatly affect their rate limit but I think it is nice to add a bit of friction to get the user to think a bit about what they are doing by setting a large limit rather than implicitly running any number of queries to get all the labels. 

Additionally, the labels are actually sorted by created at date by default when returned from the API so they are determanistic. I am open to expanding the feature set of the `label list` command to include ordering and searching as the API supports those but feels like scope creep to add in here and should be addressed in a separate PR.

Reply:
In order to get `clone` working and put off major `list` work till another time I think it would make sense to pass in `-1` or perhaps `0` into `listLabels` as the limit and that function know that means to keep retrieving until there are no more pages.

Reply:
I'll follow up after this PR with another for `label list` (and `label remove`) but I'll lay the groundwork for paging as you suggest.